### ISSUE TITLE
Allow products to be excluded from ticket-generation

### ISSUE THREAD

**Issue by pc-coholic** [pretix/pretix#456]
It would be nice to have a checkbox for every product to dis-/enable the generation of tickets (PDF, passbook, etc.).

Usecase: In addition to an entrance-ticket for an event, also offer a donation-product. When generating the downloadable tickets, the donations do not trigger a (useless) printout with a barcode which no one is going to use.

So basically we could either add a "is a downloadable ticket"-property to each product or add a "generate only admission tickets"-checkbox to the ticket generation process.

**Comment by raphaelm** [comment]
>  "generate only admission tickets"-checkbox to the ticket generation process.

I believe this is the way to go.

**Comment by hd719** [comment]
Hello I am new to Open source can I work on this?

**Comment by raphaelm** [comment]
Sure!

First of all, you need pretix running locally on your machine. We have a guide on the developer installation [over here](https://docs.pretix.eu/en/latest/development/setup.html). Play around with it and get a feeling for what pretix is and how it works :) Try to find the places in the interface mentioned in this issue where something needs to change. Maybe try to find the same places in the code? Just searching for characteristic words can be a powerful tool to find your way around.

If you have any questions, just let me know!

**Comment by hd719** [comment]
Hey @raphaelm sorry, but I cannot work on this.

**Comment by raphaelm** [comment]
Hi @hd719, sure, we'll find someone else :)

**Comment by aph3rson** [comment]
https://github.com/pretix/pretix/issues/443#issuecomment-298418044
@raphaelm I can take a crack at it. The description is a bit vague, though - what I'm getting is that you'd like products to have a property that tickets should be generated for them (so donations/etc. don't generate tickets). I'll need to look through the code a bit more to determine the program flow around when tickets are generated, though.

**Comment by raphaelm** [comment]
@aph3rson 

> The description is a bit vague, though - what I'm getting is that you'd like products to have a property that tickets should be generated for them (so donations/etc. don't generate tickets)

Nearly correct: There already *is* a property ``Item.admission`` that we want to re-use here. You would now need to:

* Add a setting to ``TicketSettingsForm`` that says "Generate tickets for non-admission products". For backwards compatibility, you should set the default to True in ``pretix/base/settings.py``.
* Disable the download button in ``pretix/presale/templates/pretixpresale/event/fragment_cart.html`` if the setting is unchecked and a product is not an admission product
* Enforce the new setting in ``OrderDownload`` (``pretix/presale/views/order.py``) when downloading a single ticket
* Add a check to ``BaseTicketOutput.generate_order`` (``pretix/base/ticketoutput.py``) and ``PdfTicketOutput.generate_order`` (``pretix/plugins/ticketoutputpdf/ticketoutput.py``) to exclude non-admission products from multi-ticket downloads

If you have any further questions here, just let me know, I'm happy to help

**Comment by aph3rson** [comment]
@raphaelm One more question, regarding PRs. The contribution guide says I need to squash my commits, though I don't see other PRs doing this. Do you prefer squashed commits for this? I generally fork master (or develop, if there is one) to an issue branch, then non-ff merge back to the appropriate branch, and submitting a PR on that (roughly following [Vincent Driessen's model](http://nvie.com/posts/a-successful-git-branching-model/)).

**Comment by raphaelm** [comment]
Hi @aph3rson, thanks for the hint, I'll drop that from the contribution guides. The guide was written before GitHub had a "Squash and merge" button, so it's easier to review if you keep your work in "natural" commits. I will then rebase/merge against master once the PR is done so we have a mostly-linear history in the repository.

Should you have a long-running PR that has merge conflicts with master once its complete, I prefer you to do a rebase over a merge of master into the issue branch
